src/main.c

```diff
@@ -42,6 +44,119 @@ static int read_image(const char *filename, png24_image *image)
     return retval;
 }
 
+
+
+#ifdef USE_LIBJPEG
+#include <jpeglib.h>
+static int read_image_jpeg(const char *filename, png24_image *image)
+{
+    int retval;
+    unsigned int width,height,row_stride;
+    unsigned int bpp;
+    unsigned int x,y,i;
+
+    struct jpeg_decompress_struct cinfo;
+    struct jpeg_error_mgr jerr;
+    cinfo.err = jpeg_std_error(&jerr);
+    jpeg_create_decompress(&cinfo);
+
+    // decompress
+    FILE *file = fopen(filename,"rb");
+    if(!file)
+    {
+        return 1;
+    }
+    jpeg_stdio_src(&cinfo,file);
+    retval=jpeg_read_header(&cinfo,TRUE);
+    if(retval != 1)
+    {
+        printf("invalid jpeg header\n");
+        return 1;
+    }
+    jpeg_start_decompress(&cinfo);
+
+    width=cinfo.output_width;
+    height=cinfo.output_height;
+    bpp=cinfo.output_components;
+    row_stride=width*bpp;
+
+    // allocate buffer size (always use RGBA)
+    unsigned char* buffer = (unsigned char*)malloc(width*height*bpp);
+    memset(buffer,0x0,width*height*bpp);
+
+    while(cinfo.output_scanline < height)
+    {
+        unsigned char *buffer_array[1];
+        buffer_array[0] = buffer + cinfo.output_scanline*row_stride;
+        jpeg_read_scanlines(&cinfo,buffer_array,1);
+    }
+
+    //convert to RGBA
+    image->rgba_data = (unsigned char*)malloc(width*height*4);
+    memset(image->rgba_data,0x0,width*height*4);
+    image->row_pointers = (unsigned char**) malloc(height*sizeof(unsigned char*));
+
+    for(y=0;y<height;y++)
+    {
+        for(x=0;x<width;x++)
+        {
+            for(i=0;i<bpp;i++)
+            {
+                image->rgba_data[(y*width*4)+(x*4)+i] = buffer[(y*width*bpp)+(x*bpp)+i];
+            }
+            // default alpha 255
+            image->rgba_data[(y*width*4) + (x*4) + 3] = 0xae;
+        }
+        image->row_pointers[y] = &image->rgba_data[y*width*4];
+    }
+    image->width=width;
+    image->height=height;
+    jpeg_destroy_decompress(&cinfo);
+    fclose(file);
+    free(buffer);
+    return 0;
+}
+#endif // #ifdef USE_LIBJPEG
+
+static int read_image(const char *filename, png24_image *image)
+{
+    int retval=1;
+    // read first 4 byte to determine filetype by magical number
+    FILE *fp = fopen(filename,"rb");
+    if(!fp)
+    {
+        return 1;
+    }
+    unsigned char *header = (unsigned char*)malloc(4*sizeof(unsigned char));
+    memset(header,0x0,4*sizeof(unsigned char));
+    if(!header)
+    {
+        return 1;
+    }
+    fread(header,sizeof(unsigned char),4,fp);
+    fclose(fp);
+
+    // the png number is not really precise but I guess the situation where this would falsely pass is almost equal to 0
+    if(header[0]==0x89 && header[1]==0x50 && header[2]==0x4e && header[3]==0x47)
+    {
+        retval=read_image_png(filename,image);
+    }
+#ifdef USE_LIBJPEG
+    else
+    if(header[0]==0xff && header[1]==0xd8 && header[2]==0xff && (header[3]==0xdb || header[3]==0xe0 || header[3]==0xe1))
```

Comment:
Checking for png is enough. Let libpng trip over anything else that isn't a JPEG.